packages/firestore/src/remote/online_state_tracker.ts

```diff
@@ -0,0 +1,166 @@
+const MAX_WATCH_STREAM_FAILURES = 2;
+
+// To deal with stream attempts that don't succeed or fail in a timely manner,
+// we have a maximum timeout we'll wait for the stream to either succeed or fail
+// MAX_WATCH_STREAM_FAILURES times, else we revert to OnlineState.Offline.
+const MAX_WATCH_STREAM_TIMEOUT_MS = 10 * 1000;
+
+/**
+ * A component used by the RemoteStore to track the OnlineState (that is,
+ * whether or not the client as a whole should be considered to be online or
+ * offline), implementing the appropriate heuristics.
+ *
+ * In particular, when the client is trying to connect to the backend, we
+ * allow up to MAX_WATCH_STREAM_FAILURES within MAX_WATCH_STREAM_TIMEOUT_MS for
+ * a connection to succeed. If we have too many failures or the timeout elapses,
+ * then we set the OnlineState to Offline, and the client will behave as if
+ * it is offline (get()s will return cached data, etc.).
+ */
+export class OnlineStateTracker {
+  /** The current OnlineState. */
+  private state = OnlineState.Unknown;
+
+  /**
+   * A count of consecutive failures to open the stream. If it reaches the
+   * maximum defined by MAX_WATCH_STREAM_FAILURES, we'll set the OnlineState to
+   * Offline.
+   */
+  private watchStreamFailures = 0;
+
+  /**
+   * A timer that elapses after MAX_WATCH_STREAM_TIMEOUT_MS, at which point we
+   * revert to OnlineState.Offline without waiting for the stream to actually
+   * fail (MAX_WATCH_STREAM_FAILURES times).
+   */
+  private watchStreamTimerPromise: CancelablePromise<void> | null = null;
+
+  /**
+   * Whether the client should log a warning message if it fails to connect to
+   * the backend (initially true, cleared after a successful stream, or if we've
+   * logged the message already).
+   */
+  private shouldWarnClientIsOffline = true;
+
+  constructor(
+    private asyncQueue: AsyncQueue,
+    private onlineStateHandler: (onlineState: OnlineState) => void
+  ) {}
+
+  /**
+   * Called by RemoteStore when a watch stream is started.
+   *
+   * It sets the OnlineState to Unknown and starts a MAX_WATCH_STREAM_TIMEOUT_MS
+   * timer if necessary.
+   */
+  handleWatchStreamStart(): void {
+    this.setAndBroadcast(OnlineState.Unknown);
+
+    if (this.watchStreamTimerPromise === null) {
+      this.watchStreamTimerPromise = this.asyncQueue.enqueueAfterDelay(
+        TimerId.OnlineStateTimeout,
+        MAX_WATCH_STREAM_TIMEOUT_MS,
+        () => {
+          this.watchStreamTimerPromise = null;
+          assert(
+            this.state === OnlineState.Unknown,
+            'Timer should be canceled if we transitioned to a different state.'
+          );
+          log.debug(
+            LOG_TAG,
+            `Watch stream didn't reach online or offline within ` +
+              `${MAX_WATCH_STREAM_TIMEOUT_MS}ms. Considering client offline.`
+          );
+          this.logClientOfflineWarningIfNecessary();
+          this.setAndBroadcast(OnlineState.Offline);
+          return Promise.resolve();
```

Comment:
Should this failure to connect in time bump the number of watch stream failures? If not could you please add comments why not?